agent/xds/listeners_ingress.go

```diff
@@ -154,21 +137,101 @@ func (s *ResourceGenerator) makeIngressGatewayListeners(address string, cfgSnap
 	return resources, nil
 }
 
-func resolveListenerSDSConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*structs.GatewayTLSSDSConfig, error) {
+func makeDownstreamTLSContextFromSnapshotListenerConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*envoy_tls_v3.DownstreamTlsContext, error) {
+	var downstreamContext *envoy_tls_v3.DownstreamTlsContext
+
+	tlsContext, err := makeCommonTLSContextFromSnapshotListenerConfig(cfgSnap, listenerCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if tlsContext != nil {
+		downstreamContext = &envoy_tls_v3.DownstreamTlsContext{
+			CommonTlsContext:         tlsContext,
+			RequireClientCertificate: &wrappers.BoolValue{Value: false},
+		}
+	}
+
+	return downstreamContext, nil
+}
+
+func makeCommonTLSContextFromSnapshotListenerConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*envoy_tls_v3.CommonTlsContext, error) {
+	var tlsContext *envoy_tls_v3.CommonTlsContext
+
+	// Enable connect TLS if it is enabled at the Gateway or specific listener
+	// level.
+	gatewayTLSCfg := cfgSnap.IngressGateway.TLSConfig
+	connectTLSEnabled := gatewayTLSCfg.Enabled ||
+		(listenerCfg.TLS != nil && listenerCfg.TLS.Enabled)
+
+	tlsCfg, err := resolveListenerTLSConfig(&gatewayTLSCfg, listenerCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if tlsCfg.SDS != nil {
```

Comment:
Same comment as above - if `tslCfg.SDS != nil` is the green path, it's better to invert this and return an error if it's ever nil

Reply:
`tlsCfg.SDS` is an optional field - in many cases it could be safely unset, falling back to serving Connect leaf certificates.